README.md
# Activate Persistence Framework

https://codeship.io/projects/6791ccc0-c291-0131-1750-5afae8528f01/status

![Activate splash](http://activate-framework.org/wp-content/uploads/2012/04/activateslider4.jpg)

Documentation ([index](/activate-docs/_index.md)):

- [Get started](/activate-docs/get-started.md)
- [Persistence Context](/activate-docs/persistence-context.md)
- [Entity](/activate-docs/entity.md) 
- [Transaction](/activate-docs/transaction.md)
- [Validation](/activate-docs/validation.md)
- [Query](/activate-docs/query.md)
- [Test support](/activate-docs/test-support.md)
- [Migration](/activate-docs/migration.md)
- [Multiple VMs](/activate-docs/multiple-vms.md)
- [Mass statement](/activate-docs/mass-statement.md)
- [Play framework](/activate-docs/play-framework.md)
- [Lift framework](/activate-docs/lift-framework.md)
- [Spray Json](/activate-docs/spray-json.md)
- [Architecture](/activate-docs/architecture.md)

## Introduction

Activate is a framework to persist objects in Scala. It is a [STM (Software Transactional Memory)](http://en.wikipedia.org/wiki/Software_transactional_memory) durable, with pluggable persistence. Its core is the RadonSTM, which provides a powerful mechanism for controlling transactions in memory, analogous to the transactions of databases, to do [optimistic concurrency control](http://en.wikipedia.org/wiki/Optimistic_concurrency_control). The durability of transactions (persistence) is pluggable and can use persistence in different paradigms such as relational ([JDBC](http://en.wikipedia.org/wiki/Java_Database_Connectivity)), prevalence ([Prevayler](http://prevayler.org/)) and non-relational ([MongoDB](https://www.mongodb.org/)).

## Benefits

The main benefits of the framework are:

- Atomic, consistent, isolated and durable transactions. You can use entities without worrying about concurrency issues.
- Entities are always consistent in memory and in the persistence layer. For example, if rollback occurs, entities in memory stay consistent.
- Transaction propagation control, including nested transactions.
- Entities are lazy loaded and initialized automatically when needed.
- Queries are type-safe and consistent, even with objects created in the current transaction. Therefore, an entity created in the same transaction may be returned in a query.
- The available memory is used efficiently, minimizing the conversation with the storage and maximizing performance.

## Build

Use sbt 0.11.2 to build Activate. Use the command "eclipse" to generate the eclipse project.
To run tests, you have to provide the databases instances for the contexts defined on the net.fwbrasil.activate.ActivateTest.contexts method of the activate-tests project.

## License

The code is licensed under [LGPL](http://pl.wikipedia.org/wiki/GNU_Lesser_General_Public_License).



CHANGELOG.md
# Changelog

## 1.6

- [finagle mysql](/activate-docs/persistence-context.md)

## 1.5

- [custom ids](/activate-docs/entity.md#custom-id)
- [mysql async](/activate-docs/persistence-context.md)
- [reads without transaction](/activate-docs/transaction.md#transaction)
- [new play integration](/activate-docs/play-framework.md#activateplaycontext)
- [test support](/activate-docs/test-support.md)
- [deferred read validation](/activate-docs/multiple-vms.md#deferred-read-validation)
- [query collection operators (in/notInt)](/activate-docs/query.md#collection-operators)
- [custom classpath scan](/activate-docs/persistence-context.md)
- [persisted index](/activate-docs/persistence-context.md#persisted-index)
- [zip encoder](/activate-docs/entity.md#custom-type-encoders)
- [bug fixes](https://github.com/fwbrasil/activate/issues?milestone=4&page=1&state=closed)

## 1.4

- Prevalent storage
- Slick 2 lifted embedding support
- Sql Server dialect
- Custom type encoders
- Lift Framewok integration
- Update to Play 2.2
- Custom cache configuration
- Eager queries
- Cached queries
- Dynamic queries
- Memory index
- Lifecycle listeners
- Property listeners
- EntityMap
- Cascade delete
- Kryo serializer
- Modify column type migration
- Queries with empty where

## 1.3

- Async transactions
- Async queries
- Reactive PostgreSQL storage (using postgresql-async)
- Reactive MongoDB storage (using reactivemongo)
- Slick direct embedding queries
- Spray json integration
- Lazy entity lists
- Limited queries offset
- toUpperCase/toLowerCase query functions

## 1.2

- Scala 2.10
- Polyglot persistence
- Optimistic offline locking
- Play 2.1 support
- DB2 jdbc dialect
- Limited queries
- Better transient field support
- Custom serializers
- Nested list order preservation


## 1.1

- Play application hot reload support (ActivatePlayPlugin)
- New query syntax option: select[MyEntity] where(_.name :== “a”)
- New databases: h2, derby and hsqldb.
- Support for list fields in entities: class MyEntity(var intList: List[Int]) extends Entity
- @Alias annotation to customize entity and field names at the storage
- Manual migrations
- Multiple VMs support (Coordinator)
- Performance enhancements

## 1.0

- Migrations
- Play Framework support
- Mass update/delete statements
- Validation using Design By Contract
- Queries with “order by”
- Paginated queries
- Storage direct access
- System properties defined storage
- Better error messages
- It’s not necessary to use “-noverify” anymore
- Performance enhancements


project/ActivateBuild.scala
import sbt._
import Keys._

object ActivateBuild extends Build {

    /* Core dependencies */
    val javassist = "org.javassist" % "javassist" % "3.18.1-GA"
    val radonStm = "net.fwbrasil" %% "radon-stm" % "1.6"
    val smirror = "net.fwbrasil" %% "smirror" % "0.8"
    val guava = "com.google.guava" % "guava" % "16.0"
    val objenesis = "org.objenesis" % "objenesis" % "2.1"
    val jug = "com.fasterxml.uuid" % "java-uuid-generator" % "3.1.3"
    val reflections = "org.reflections" % "reflections" % "0.9.8" exclude ("javassist", "javassist") exclude ("dom4j", "dom4j")
    val grizzled = "org.clapper" %% "grizzled-slf4j" % "1.0.1"
    val logbackClassic = "ch.qos.logback" % "logback-classic" % "1.1.0"
    val jodaTime = "joda-time" % "joda-time" % "2.3"
    val jodaConvert = "org.joda" % "joda-convert" % "1.6"
    val blueprintsCore = "com.tinkerpop.blueprints" % "blueprints-core" % "2.4.0"
    val blueprintsNeo4j = "com.tinkerpop.blueprints" % "blueprints-neo4j-graph" % "2.4.0"
    val gremlin = "com.tinkerpop.gremlin" % "gremlin-java" % "2.4.0"
    val xstream = "com.thoughtworks.xstream" % "xstream" % "1.4.6" exclude ("xpp3", "xpp3_min")
    val jettison = "org.codehaus.jettison" % "jettison" % "1.3.4"
    val findBugs = "com.google.code.findbugs" % "jsr305" % "2.0.1"
    val kryo = "com.esotericsoftware.kryo" % "kryo" % "2.23.0"
    val cassandraDriver = "com.datastax.cassandra" % "cassandra-driver-core" % "1.0.5"

    /* Prevayler */
    val prevaylerCore = "org.prevayler" % "prevayler-core" % "2.6"
    val prevaylerFactory = "org.prevayler" % "prevayler-factory" % "2.6"
    val prevaylerXStream = "org.prevayler.extras" % "prevayler-xstream" % "2.6"

    /* 
    Install oracle in your local repo
  */
    val objbd6 = "com.oracle" % "ojdbc6" % "11.2.0"
    val mysql = "mysql" % "mysql-connector-java" % "5.1.28"
    val postgresql = "org.postgresql" % "postgresql" % "9.3-1100-jdbc4"
    val hirakiCP = "com.zaxxer" % "HikariCP" % "1.3.8"
    val h2 = "com.h2database" % "h2" % "1.3.175"
    val derby = "org.apache.derby" % "derby" % "10.10.1.1"
    val hqsqldb = "org.hsqldb" % "hsqldb" % "2.3.1"
    val db2jcc = "com.ibm.db2.jcc" % "db2jcc4" % "10.1"
    val jtds = "net.sourceforge.jtds" % "jtds" % "1.3.1"

    val gfork = "org.gfork" % "gfork" % "0.11"

    /* Mongo */
    val mongoDriver = "org.mongodb" % "mongo-java-driver" % "2.11.4"

    lazy val activate =
        Project(
            id = "activate",
            base = file("."),
            aggregate = Seq(activateCore, activatePrevayler,
                activateJdbc, activateMongo, activateTest, activatePlay,
                activateGraph, activateSprayJson, activateJdbcAsync,
                activateSlick, activateMongoAsync, activatePrevalent,
                activateCassandraAsync, activateLift, activateFinagleMysql),
            settings = commonSettings
        )

    lazy val activateCore =
        Project(
            id = "activate-core",
            base = file("activate-core"),
            settings = commonSettings ++ Seq(
                libraryDependencies ++=
                    Seq(javassist, radonStm, objenesis, jug,
                        reflections, grizzled, logbackClassic, jodaTime, jodaConvert,
                        smirror, xstream, jettison, findBugs, kryo, guava)
            )
        )

    lazy val activatePrevayler =
        Project(
            id = "activate-prevayler",
            base = file("activate-prevayler"),
            dependencies = Seq(activateCore),
            settings = commonSettings ++ Seq(
                libraryDependencies ++=
                    Seq(prevaylerCore, prevaylerFactory, prevaylerXStream)
            )
        )

    lazy val activatePrevalent =
        Project(
            id = "activate-prevalent",
            base = file("activate-prevalent"),
            dependencies = Seq(activateCore),
            settings = commonSettings
        )

    lazy val activateCassandraAsync =
        Project(
            id = "activate-cassandra-async",
            base = file("activate-cassandra-async"),
            dependencies = Seq(activateCore),
            settings = commonSettings ++ Seq(
                libraryDependencies ++=
                    Seq(cassandraDriver)
            )
        )

    lazy val activateJdbc =
        Project(
            id = "activate-jdbc",
            base = file("activate-jdbc"),
            dependencies = Seq(activateCore),
            settings = commonSettings ++ Seq(
                libraryDependencies ++=
                    Seq(hirakiCP)
            )
        )

    val slick = "com.typesafe.slick" % "slick_2.10" % "2.0.0"
    val scalaCompiler = "org.scala-lang" % "scala-compiler" % "2.10.3"

    lazy val activateSlick =
        Project(
            id = "activate-slick",
            base = file("activate-slick"),
            dependencies = Seq(activateCore, activateJdbc),
            settings = commonSettings ++ Seq(
                libraryDependencies ++=
                    Seq(scalaCompiler, slick)
            )
        )

    val postgresqlAsync = "com.github.mauricio" %% "postgresql-async" % "0.2.13"
    val mysqlAsync = "com.github.mauricio" %% "mysql-async" % "0.2.13"

    lazy val activateJdbcAsync =
        Project(
            id = "activate-jdbc-async",
            base = file("activate-jdbc-async"),
            dependencies = Seq(activateCore, activateJdbc),
            settings = commonSettings ++ Seq(
                libraryDependencies ++=
                    Seq(postgresqlAsync, mysqlAsync)
            )
        )

    val finagleMysql = "com.twitter" %% "finagle-mysql" % "6.16.0"

    lazy val activateFinagleMysql =
        Project(
            id = "activate-finagle-mysql",
            base = file("activate-finagle-mysql"),
            dependencies = Seq(activateCore, activateJdbc),
            settings = commonSettings ++ Seq(
                libraryDependencies ++=
                    Seq(finagleMysql)
            )
        )

    lazy val activateMongo =
        Project(
            id = "activate-mongo",
            base = file("activate-mongo"),
            dependencies = Seq(activateCore),
            settings = commonSettings ++ Seq(
                libraryDependencies ++=
                    Seq(mongoDriver)
            )
        )

    val reactivemongo = "org.reactivemongo" %% "reactivemongo" % "0.10.0"

    lazy val activateMongoAsync =
        Project(
            id = "activate-mongo-async",
            base = file("activate-mongo-async"),
            dependencies = Seq(activateCore, activateMongo),
            settings = commonSettings ++ Seq(
                libraryDependencies ++=
                    Seq(reactivemongo)
            )
        )

    lazy val activateGraph =
        Project(
            id = "activate-graph",
            base = file("activate-graph"),
            dependencies = Seq(activateCore),
            settings = commonSettings ++ Seq(
                libraryDependencies ++=
                    Seq(blueprintsCore, gremlin)
            )
        )

    val play = "com.typesafe.play" %% "play" % "2.2.1"
    val playTest = "com.typesafe.play" %% "play-test" % "2.2.1"

    lazy val activatePlay =
        Project(
            id = "activate-play",
            base = file("activate-play"),
            dependencies = Seq(activateCore, activateTest, activateJdbc),
            settings = commonSettings ++ Seq(
                libraryDependencies ++=
                    Seq(play, playTest % "provided", specs2 % "provided")
            )
        )

    val lift = "net.liftweb" %% "lift-webkit" % "2.5.1"

    lazy val activateLift =
        Project(
            id = "activate-lift",
            base = file("activate-lift"),
            dependencies = Seq(activateCore),
            settings = commonSettings ++ Seq(
                libraryDependencies ++=
                    Seq(lift)
            )
        )

    val sprayJson = "io.spray" %% "spray-json" % "1.2.5"

    lazy val activateSprayJson =
        Project(
            id = "activate-spray-json",
            base = file("activate-spray-json"),
            dependencies = Seq(activateCore),
            settings = commonSettings ++ Seq(
                libraryDependencies ++=
                    Seq(sprayJson)
            )
        )

    val jackson = Seq(
        "com.fasterxml.jackson.core" % "jackson-core" % "2.3.1",
        "com.fasterxml.jackson.core" % "jackson-annotations" % "2.3.1",
        "com.fasterxml.jackson.core" % "jackson-databind" % "2.3.1",
        "com.fasterxml.jackson.module" %% "jackson-module-scala" % "2.3.1")

    lazy val activateJacksonJson =
        Project(
            id = "activate-jackson",
            base = file("activate-jackson"),
            dependencies = Seq(activateCore, activateJdbc),
            settings = commonSettings ++ Seq(
                libraryDependencies ++=
                    Seq(postgresql, scalaCompiler) ++ jackson
            )
        )

    val junit = "junit" % "junit" % "4.11"
    val specs2 = "org.specs2" %% "specs2" % "2.3.7"

    lazy val activateTest =
        Project(id = "activate-test",
            base = file("activate-test"),
            dependencies = Seq(activateCore, activatePrevayler % "test", activateJdbc % "test",
                activateMongo % "test", activateGraph % "test", activateSprayJson % "test", activateJacksonJson % "test", activateJdbcAsync % "test",
                activateSlick % "test", activateMongoAsync % "test", activatePrevalent % "test", activateCassandraAsync % "test", activateLift % "test",
                activateFinagleMysql % "test"),
            settings = commonSettings ++ Seq(
                libraryDependencies ++=
                    Seq(junit % "test", specs2 % "test", mysql % "test", objbd6 % "test", postgresql % "test", db2jcc % "test",
                        h2 % "test", derby % "test", hqsqldb % "test", gfork % "test", blueprintsNeo4j % "test", jtds % "test"),
                scalacOptions ++= Seq("-Xcheckinit")
            )
        )

    /* Resolvers */
    val customResolvers = Seq(
        "Maven" at "http://repo1.maven.org/maven2/",
        "Typesafe" at "http://repo.typesafe.com/typesafe/releases",
        "Local Maven Repository" at "file://" + Path.userHome + "/.m2/repository",
        "fwbrasil.net" at "http://fwbrasil.net/maven/",
        "spray" at "http://repo.spray.io/",
        "Sonatype OSS Releases" at "http://oss.sonatype.org/content/repositories/releases/",
        "Sonatype OSS Snapshots" at "http://oss.sonatype.org/content/repositories/snapshots/",
        "Alfesco" at "https://maven.alfresco.com/nexus/content/groups/public/"
    )

    def commonSettings =
        Defaults.defaultSettings ++ Seq(
            organization := "net.fwbrasil",
            version := "1.6",
            scalaVersion := "2.10.3",
            javacOptions ++= Seq("-source", "1.7", "-target", "1.7"),
            publishMavenStyle := true,
            // publishTo := Some(Resolver.file("file",  new File(Path.userHome.absolutePath+"/.m2/repository"))), 
            // publishTo := Option(Resolver.ssh("fwbrasil.net repo", "fwbrasil.net", 8080) as("maven") withPermissions("0644")),
            publishTo <<= version { v: String =>
                val nexus = "https://oss.sonatype.org/"
                val fwbrasil = "http://fwbrasil.net/maven/"
                if (v.trim.endsWith("SNAPSHOT"))
                    Option(Resolver.ssh("fwbrasil.net repo", "fwbrasil.net", 8080) as ("maven") withPermissions ("0644"))
                else
                    Some("releases" at nexus + "service/local/staging/deploy/maven2")
            },
            resolvers ++= customResolvers,
            publishMavenStyle := true,
            publishArtifact in Test := false,
            pomIncludeRepository := { x => false },
            pomExtra := (
                <url>http://github.com/fwbrasil/activate/</url>
                <licenses>
                    <license>
                        <name>LGPL</name>
                        <url>https://github.com/fwbrasil/activate/blob/master/LICENSE-LGPL</url>
                        <distribution>repo</distribution>
                    </license>
                </licenses>
                <scm>
                    <url>git@github.com:fwbrasil/activate.git</url>
                    <connection>scm:git:git@github.com:fwbrasil/activate.git</connection>
                </scm>
                <developers>
                    <developer>
                        <id>fwbrasil</id>
                        <name>Flavio W. Brasil</name>
                        <url>http://fwbrasil.net</url>
                    </developer>
                </developers>
            ),
            compileOrder := CompileOrder.JavaThenScala,
            parallelExecution in Test := false
        )
}
